src/dnstest_harness.erl

```
@@ -220,16 +365,60 @@ test_header(ExpectedHeader, Response) ->
     Packet = dns:encode_message(Message),
-    ?LOG_DEBUG("Sending UDP query to host ~p and port ~p", [host(), port()]),
-    {ok, Socket} = gen_udp:open(0, [binary, {active, false}]),
-    ok = gen_udp:send(Socket, Host, Port, Packet),
-    QueryResponse =
-        case gen_udp:recv(Socket, 65535, 6000) of
-            {ok, {Host, _Port, Reply}} when is_binary(Reply) -> {ok, dns:decode_message(Reply)};
-            {error, Error} -> {error, {Error, {server, {Host, Port}}}}
-        end,
-    gen_udp:close(Socket),
-    QueryResponse.
+    ?LOG_DEBUG("Sending UDP query to host ~p and port ~p", [Host, Port]),
+    case gen_udp:open(0, [binary, {active, false}]) of
+        {ok, Socket} ->
+            try
+                ok = gen_udp:send(Socket, Host, Port, Packet),
+                % Increased buffer size for potential large UDP responses
+                case gen_udp:recv(Socket, 65535, 6000) of
+                    {ok, {_RHost, _RPort, Reply}} when is_binary(Reply) ->
+                        {ok, dns:decode_message(Reply)};
+                    {error, timeout} ->
+                        ?LOG_WARNING("UDP query timed out for ~p", [Message#dns_message.questions]),
+                        {error, {timeout, {server, {Host, Port}}}};
+                    {error, Error} ->
+                        ?LOG_ERROR("UDP recv error: ~p", [Error]),
+                        {error, {Error, {server, {Host, Port}}}}
+                end
+            after
+                gen_udp:close(Socket)
+            end;
+        {error, Reason} ->
+            ?LOG_ERROR("Failed to open UDP socket: ~p", [Reason]),
+            {error, {socket_error, Reason}}
+    end.
+
+% Send the message to the given host:port via TCP.
+send_tcp_query(Message, Host, Port) ->
+    Packet = dns:encode_message(Message),
+    ?LOG_DEBUG("Sending TCP query to host ~p and port ~p", [Host, Port]),
+    case gen_tcp:connect(Host, Port, [binary, {packet, 2}, {active, false}], 6000) of
```

Any reason why `{packet, 2}` and not `{packet, raw}`? There's barely an occasion to use that, I'm curious, does this work correctly when decoding the received packet? 🤔

Timeout can also be something much shorter, maybe 1s max, this is just connecting the socket and if this takes that long in tests we probably want to fail fast.

---

[RFC 1035 Section 4.2.2](https://www.rfc-editor.org/rfc/rfc1035#section-4.2.2)
> Messages sent over TCP connections use server port 53 (decimal).  The
message is prefixed with a two byte length field which gives the message

---

wow, till I knew 😳

---

TImeout set to 1s [00cefa6](https://github.com/dnsimple/dnstest/pull/21/commits/00cefa67a45a1084e16a346aa57145cbd4ee064c)